causalml/propensity.py

```diff
@@ -1,46 +1,32 @@
+from abc import ABCMeta, abstractproperty
 import logging
 import numpy as np
 from pygam import LogisticGAM, s
 from sklearn.metrics import roc_auc_score as auc
 from sklearn.linear_model import LogisticRegressionCV
 from sklearn.model_selection import StratifiedKFold, train_test_split
 import xgboost as xgb
-import multiprocessing
 
-logger = logging.getLogger('causalml')
 
+logger = logging.getLogger('causalml')
 
-class ElasticNetPropensityModel(object):
-    """Propensity regression model based on the ElasticNet algorithm.
 
-    Attributes:
-        model (sklearn.linear_model.ElasticNetCV): a propensity model object
-    """
-
-    def __init__(self, n_fold=4, Cs=np.logspace(1e-3, 1 - 1e-3, 4), l1_ratios=np.linspace(1e-3, 1 - 1e-3, 4),
-                 clip_bounds=(1e-3, 1 - 1e-3), cv=None, random_state=None):
-        """Initialize a propensity model object.
+class PropensityModel(metaclass=ABCMeta):
+    def __init__(self, clip_bounds=(1e-3, 1 - 1e-3), **model_kwargs):
+        """
         Args:
-            n_fold (int): the number of cross-validation fold
-            Cs (int or array-like): Each of the values in Cs describes the inverse of regularization strength.
-                If Cs is as an int, then a grid of Cs values are chosen in a logarithmic scale between 1e-4 and 1e4.
-            l1_ratios (array-like): array of l1 ratios (0 <= value <= 1) to iterate through in CV estimator
-            clip_bounds (tuple): lower and upper bounds for clipping propensity scores. Bounds should be implemented
-                such that: 0 < lower < upper < 1, to avoid division by zero in BaseRLearner.fit_predict() step.
-            cv (int or cross-validation generator): The default cross-validation generator used is Stratified K-Folds.
-                If an integer is provided, then it is the number of folds used.
-            random_state (numpy.random.RandomState or int): RandomState or an int seed
 
-        Returns:
-            None
+        clip_bounds (tuple): lower and upper bounds for clipping propensity scores. Bounds should be implemented
+                such that: 0 < lower < upper < 1, to avoid division by zero in BaseRLearner.fit_predict() step.
+        model_kwargs: Keyword arguments to be passed to the underlying classification model.
         """
-        if cv is None:
-            self.cv = StratifiedKFold(n_splits=n_fold, shuffle=True, random_state=random_state)
-        else:
-            self.cv = cv
-        self.model = LogisticRegressionCV(penalty='elasticnet', solver='saga', Cs=Cs, l1_ratios=l1_ratios,
-                                          cv=self.cv, random_state=random_state)
         self.clip_bounds = clip_bounds
+        self.model_kwargs = model_kwargs
+        self.model = self._model
+
+    @abstractproperty
```

Comment:
`@abstractproperty` is deprecated since Python 3.3. Please use `@property` and `@abstractmethod` as suggested in the [doc](https://docs.python.org/3/library/abc.html#abc.abstractproperty).

Reply:
Learned something new. Thank you.